### DISCUSSION TITLE
How to know if any columns in the ActiveModel are in a Set state.

### DISCUSSION THREAD

**Discussion by nickb937** [SeaQL/sea-orm#502] [Q&A] [answered]
At the moment, it is possible to create an UPDATE statement where the SET field is empty i.e. `UPDATE table SET WHERE id=1`

```
let row = table::Entity::find_by_id(1).one(dbh).await?.unwrap();
let mut arow: table::ActiveModel = row.into_active_model();
arow.update(dbh).await?;
```

I'd like to be able to at least inspect the ActiveModel to know if all columns are unmodified and therefore not call `update()` e.g. a method like  `arow.are_any_set()` which could either be a `Vec<table::Column>` or a simple `bool`.  Does anybody know if there is an existing way to do this?

**Comment by billy1624** [comment]
Hey @nickb937, sorry for the late reply. There is a [`ActiveValue::is_set`](https://docs.rs/sea-orm/0.6.0/sea_orm/entity/enum.ActiveValue.html#method.is_set) method.

For example, you can check if `id` is set by
```rust
assert!(arow.id.is_set())
```

**Comment by nickb937** [comment]
The `is_set()` acts on the singular column, but I couldn't use this to validate that all columns in the ActiveModel were unchanged in a future-proof way.  For example at time `T`, the Model definition might be:

```rust
struct Table {
  A: i32,
  B: i32
}
```

At time `T1`, the Model definition might be:

```rust
struct Table {
  A: i32,
  B: i32,
  C: i32
}
```

Consider if the `Table::ActiveModel` is created in function `foo()`, mutated in function `bar()` and updated to the DB in function `baz()`

At development time `T`, when `baz()` was initially written, it will test for `A.is_set() || B.is_set()`.

At time `T1` when the model is changed to add the column `C`, the developer now needs to remember this logic check in `baz()` and change it to `A.is_set() || B.is_set() || C.is_set()`.

**Reply by billy1624** [reply]
Valid concern! Try this, iterate over all columns at any time.

```rust
let active_model = fruit::ActiveModel {
    id: NotSet,
    name: NotSet,
    cake_id: Set(Some(1)),
};

assert!(fruit::Column::iter().all(|col| active_model.get(col).is_set()));
```

**Comment by nickb937** [comment, marked as answer]
Ah thanks.  I hadn't noticed there was a iter() on the Column.

**Reply by billy1624** [reply]
Hidden features loll

**Reply by JonasCir** [reply]
This should be documented then, seems like a super useful one :D

**Reply by billy1624** [reply]
Where should we put it on? Do you have any thoughts? hahaa

**Reply by JonasCir** [reply]
Currently there is an `Advanced Queries` chapter in the docs, maybe introduce an `Advanced Features` chapter with a `Column introspection` section?

**Reply by billy1624** [reply]
Yeah, it could! Let me mark it down.
Thanks!

**Reply by JonasCir** [reply]
Sure :) Thanks for all the great work and effort!
